### PR TITLE
HPCC-18422 Add new format meta data to indexes

### PR DESCRIPTION
**Pull request by richardkchapman**
Fixed some issues spotted while testting previous fix.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Ran dumpkey tests, having observed in debugger how some cases were previously working just by luck.

### CODE REVIEW

File: rtl/eclrtl/rtldynfield.cpp [hunk 1]
```diff
@@ -1025,7 +1025,7 @@ class GeneralRecordTranslator : public CInterfaceOf<IDynamicTransform>
                 size_t sourceOffset = sourceRow.getOffset(matchField);
                 const byte *source = sourceRec + sourceOffset;
                 size_t copySize = sourceRow.getSize(matchField);
-                if (copySize == 0)  // Field is missing because of an ifblock - use default value
+                if (copySize == 0 && match.matchType & match_inifblock)  // Field is missing because of an ifblock - use default value
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: cleaner with brackets around the &.  Even better if it was match.inIfBlock()

File: tools/dumpkey/dumpkey.cpp [hunk 2]
```diff
@@ -186,26 +187,46 @@ int main(int argc, const char **argv)
...
-                        size32_t minRecSize = 0;
                         ForEachItemIn(idx, fieldNames)
                         {
                             unsigned fieldNum = inrec.getFieldNum(fieldNames.item(idx));
                             if (fieldNum == (unsigned) -1)
                                 throw MakeStringException(0, "Requested output field '%s' not found", fieldNames.item(idx));
-                            fields.append(inrec.queryOriginalField(fieldNum));
-                            minRecSize += inrec.queryType(fieldNum)->getMinSize();
+                            const RtlFieldInfo *sourceField = inrec.queryOriginalField(fieldNum);
+                            if (sourceField->type->getType() == type_filepos)
```
Comment (ghalliday) [inline tools/dumpkey/dumpkey.cpp]: Could do with some comments explaining what it is doing - it isn't obvious from the code.

File: tools/dumpkey/dumpkey.cpp [hunk 3]
```diff
@@ -186,26 +187,46 @@ int main(int argc, const char **argv)
...
-                        outRecType = new RtlRecordTypeInfo(type_record, minRecSize, fields.getArray(0));
-                        outmeta.setown(new CDynamicOutputMetaData(*outRecType));
                     }
                     else
+                    {
+                        unsigned numFields = inrec.getNumFields();
+                        for (unsigned idx = 0; idx < numFields;idx++)
+                        {
+                            const RtlFieldInfo *sourceField = inrec.queryOriginalField(idx);
+                            if (sourceField->type->getType() == type_filepos)
```
Comment (ghalliday) [inline tools/dumpkey/dumpkey.cpp]: same here

**Comment by ghalliday** [comment]
@richardkchapman a couple of comments/questions

**Comment by richardkchapman** [comment]
@ghalliday Repushed

**Pull request merged**